aten/src/ATen/native/cuda/SoftMax.cu

```diff
@@ -939,18 +939,26 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
         "expected input and grad types to match, or input to be at::Half and grad to be at::Float");
   }
   Tensor tmp = grad * output;
-  host_softmax_backward<SoftMaxBackwardEpilogue,false>(tmp, output, dim, half_to_float, grad_input);
+  host_softmax_backward<SoftMaxBackwardEpilogue, false>(tmp, output, dim, half_to_float, grad_input);
 }
 
-Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
-    TORCH_CHECK(mask.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
-    bool is_transformer_mask = (input.dim() == 4 && mask.dim() == 2 && input.size(0) == mask.size(0) && input.size(2) == mask.size(1) && input.size(3) == mask.size(1));
-    TORCH_CHECK(mask.sizes() == input.sizes() || is_transformer_mask, "Mask shape should match input");
-    // Always do masked softmax on last dim
-    int softmax_elements = input.size(input.dim() - 1);
+Tensor masked_softmax_cuda(const Tensor& input_, int64_t dim, const Tensor& mask_) {
+    // TORCH_CHECK(mask.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
+    // bool is_transformer_mask = (input_.dim() == 4 && mask.dim() == 2 && input_.size(0) == mask.size(0) && input_.size(2) == mask.size(1) && input_.size(3) == mask.size(1));
+    // TORCH_CHECK(mask.sizes() == input_.sizes() || is_transformer_mask, "Mask shape should match input");
+
+    TORCH_CHECK(mask_.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
+    bool is_transformer_mask = (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
+    TORCH_CHECK(mask_.sizes() == input_.sizes() || is_transformer_mask, "Mask shape should match input");
+
+    auto input = input_.dim() == 0 ? input_.view(1) : input_;
+    auto mask = mask_.dim() == 0 ? mask_.view(1) : mask_;
+
+    Tensor output = at::empty_like(input, input.options());
+    int softmax_elements = input.size(dim);
+
     // Persistent softmax only support softmax_elements <= 1024,
     // Therefore once softmax_elements > 1024, we need to fallback to vanilla masked_softmax
-    Tensor output = at::empty_like(input, input.options());
     // Fallback to a slower masked softmax solution
     if (softmax_elements > 1024 || softmax_elements * input.element_size() > 4096 || !mask.is_contiguous()) {
```

Comment:
when `dim` is not the last, you also should take this fallback

Reply:
ah yes, good catch! I realized this today while debugging it as well